### PR TITLE
Implement to/fromBits in TS, and set max length to 254 bits

### PR DESCRIPTION
**Pull request by jackryanservia**
Closes: #1023

* Implements `toBits` and `fromBits` methods in TS.
* Sets length limit 254bits for both to prevent issue outlined in #1023

### CODE REVIEW

File: src/lib/field.ts
Comment (mitschabaude) [inline src/lib/field.ts:931]: we can't do `toBoolean()` in circuit - but the constraint below should throw an error in exactly the same case? so maybe put this error in the `assertEquals` message below?

**Review by jackryanservia — COMMENTED** [review]
I will think more about the `to/fromBits()` implementation and check the constraint count before I merge, but for now I just have a few questions about other things to make sure I'm on the right track! :)

File: src/lib/field.ts
Comment (jackryanservia) [inline src/lib/field.ts:938]: Is `Provable.asProver()` the correct way to log this?

**Review by jackryanservia — COMMENTED** [review]
Ok, it looks like the number of rows for both methods is exactly the same as before (and the digest is identical for `fromBits()`).

Let me know if I should add vk regression test for these methods; aside from that I think this is ready to merge if looks ok to reviewer :)

File: src/lib/nullifier.ts
Comment (jackryanservia) [reply src/lib/nullifier.ts:76]: Never mind all of this! I am still curious about the shiftiness, but now that I look at the `Scalar` code it's seems like there's nothing wrong with just having `Scalar.fromBits()` pad the end of any input less than 255 bits with `new Bool(false)`.

File: src/lib/field.ts
Comment (mitschabaude) [reply src/lib/field.ts:921]: we could provide a secure, 255-bits version by treating the bits as a `ForeignField` element and asserting that it's less than the field size. That would add some overhead so it's better to not do it by default. It could be a separate method `toFullBits()` or similar. But not urgent IMO

File: src/lib/nullifier.ts
Comment (mitschabaude) [reply src/lib/nullifier.ts]: yes agreed

Comment (mitschabaude) [reply src/lib/nullifier.ts]: `Scalar.from()` doesn't take a `Field`. It takes a `bigint` but we can't move out and in of provable types like that

File: src/lib/field.ts
Comment (mitschabaude) [reply src/lib/field.ts:938]: what lol no, this will log `undefined` because `asProver()` doesn't return anything.
I would leave out the value,
```suggestion
      `Field.toBits(): Input does not fit in ${length} bits`
```

File: src/lib/field.ts
Comment (mitschabaude) [inline src/lib/field.ts:970]: could you add `.seal()` at the end of the returned expression? that way we return a fresh variable instead of an AST of 254 summands. the latter could create an enourmous amount of unnecessary constraints when used twice in different places. see https://github.com/o1-labs/o1js/issues/1313 for more background

File: tests/vk-regression/vk-regression.json
Comment (mitschabaude) [inline tests/vk-regression/vk-regression.json:57]: 1-row reduction makes sense because the hello-world example does `privateKey.toPublicKey()` which as the final step converts the `Group` (uncompressed EC point `(x,y)`) into a `PublicKey` (compressed EC point `(x, isOdd)`) and to do that it needs to get all of `y`'s bits with `toBits()`. `toBits()` needs 1 row less now because it adds up 254 bits, not 255. 

It does mean that certain public keys are no longer supported by this method. I think we should address that separately by introducing a dedicated `parity()` gadget using range checks, which makes it more efficient as well

**Review by mitschabaude — COMMENTED** [review]
Looks great, just a few comments that should be addressed

Also @jackryanservia please add a changelog! Breaking change

File: src/lib/field.ts
Comment (jackryanservia) [reply src/lib/field.ts]: Ahhh, very interesting! :)

File: src/lib/field.ts
Comment (jackryanservia) [reply src/lib/field.ts:921]: This makes sense. I will implement tomorrow if there is time.

File: tests/vk-regression/vk-regression.json
Comment (jackryanservia) [reply tests/vk-regression/vk-regression.json:57]: I had wondered about this. Would it make sense to have this use `toFullBits()` here in the meantime?

Comment (mitschabaude) [reply tests/vk-regression/vk-regression.json:57]: Yes that would make sense

**Review by mitschabaude — APPROVED** [review]

**Review by ymekuria — APPROVED** [review]
Nice work @jackryanservia !

**Pull request merged**
